compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFuseConsumerIntoLoop.cpp

```diff
@@ -49,72 +46,74 @@ void AMDAIEFuseConsumerIntoLoopPass::runOnOperation() {
-            scfLoopOp = op;
-            return WalkResult::interrupt();
-          }
-          return WalkResult::advance();
-        });
-
-    if (!scfLoopOp) {
-      LLVM_DEBUG(llvm::dbgs()
-                 << "There is no scf.for/forall loop to fuse with\n");
-      return;
-    }
-
-    // Search the compute op and its consumer slices.
-    linalg::LinalgOp linalgOp;
-    scfLoopOp->walk<WalkOrder::PostOrder, ReverseIterator>(
-        [&](linalg::LinalgOp op) {
-          linalgOp = op;
+  // Step 2. Find the first scf loop in postorder walk.
+  Operation *scfLoopOp = nullptr;
+  funcOp->walk<WalkOrder::PostOrder, ReverseIterator>(
+      [&](LoopLikeOpInterface op) {
+        if (isa<scf::ForOp>(op)) {
+          scfLoopOp = op;
           return WalkResult::interrupt();
-        });
-
-    if (!linalgOp) {
-      LLVM_DEBUG(llvm::dbgs() << "Could not find any compute op\n");
-      return;
-    }
-
-    Value::user_range users = linalgOp->getResult(0).getUsers();
-    if (!llvm::hasSingleElement(users)) {
-      linalgOp->emitOpError("Expected only one user of the compute op");
-      return signalPassFailure();
-    }
+        } else if (isa<scf::ForallOp>(op)) {
+          scfLoopOp = op;
+          return WalkResult::interrupt();
+        }
+        return WalkResult::advance();
+      });
+  if (!scfLoopOp) {
+    LLVM_DEBUG(llvm::dbgs()
+               << "There is no scf.for/forall loop to fuse with\n");
+    return;
+  }
 
-    Operation *terminatorStoreOp = *(users.begin());
-    if (!(isa<tensor::InsertSliceOp, tensor::ParallelInsertSliceOp>(
-            terminatorStoreOp))) {
-      terminatorStoreOp->emitOpError(
-          "Expected either tensor.insert_slice OR tensor.parallel_insert_slice "
-          "to be the only user of the compute op");
-      return signalPassFailure();
-    }
+  // Step 3. Search the compute op and its consumer slices.
+  linalg::LinalgOp linalgOp;
+  scfLoopOp->walk<WalkOrder::PostOrder, ReverseIterator>(
+      [&](linalg::LinalgOp op) {
+        linalgOp = op;
+        return WalkResult::interrupt();
+      });
+  if (!linalgOp) {
+    LLVM_DEBUG(llvm::dbgs() << "Could not find any compute op\n");
+    return;
+  }
 
-    std::optional<scf::SCFFuseConsumerOfSliceResult> fusedConsumer =
-        scf::tileAndFuseConsumerOfSlice(rewriter, terminatorStoreOp);
-    if (!fusedConsumer) {
-      terminatorStoreOp->emitOpError(
-          "Failed to fuse any consumer op into the producer");
-      return signalPassFailure();
-    }
-    fusedConsumer->origConsumerOperand->getOwner()->erase();
+  Operation *computeOp = linalgOp;
+  // Step 4. Based on the `fuseDepth`, we would greedily fuse the consumer ops.
+  for (unsigned depth = 1; depth <= fuseDepth; depth++) {
+    do {
+      Value::user_range users = computeOp->getResult(0).getUsers();
+      if (!llvm::hasSingleElement(users)) {
+        computeOp->emitOpError("Expected only one user of the compute op");
```

Comment:
I guess for both pure Matmul as well as Matmul + Elemwise, we always expect this pass to succeed.

A silent bailing on an inherent (and strong) assumption that this pass makes might not be a good idea - I believe all kind of input dispatches generated via IREE would maintain the "one user of the compute op" and as such it might be worth it to raise an error if any of the preceding transformations we do in our pipeline changes that in any way.

I don't mind deleting this (and the corresponding lit tests) - let me know your thoughts.

Reply:
> A silent bailing on an inherent (and strong) assumption that this pass makes might not be a good idea

What's the inherent strong assumption? If we bail here, we're just not doing any fusion, right? With this error you're basically throwing an error on a potentially correct input we haven't seen before, assuming it will be incorrect, but which will not break this pass afaict, so I don't think this is the right place to throw an error.